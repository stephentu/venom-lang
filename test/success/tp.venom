class tuple{A, B}
  attr first::A
  attr second::B
  def self(lhs::A, rhs::B) =
    first = lhs;
    second = rhs;
  end
  def getFirst() -> A = self.first; end
  def getSecond() -> B = self.second; end
end

def printTuple{A, B}(t::tuple{A, B}) =
  print(t.getFirst());
  print(t.getSecond());
end

tupleVar::tuple{int, string} = Nil;

class s <- tuple{int, string}
  def self(a::int, b::string) : super(a, b) = end
  def addAndGet1(x::int) -> int = 
    return first + x;
  end
  def concatAndGet2(x::string) -> string = 
    return second;
  end
end

obj = s(10, "foo");
print(obj.addAndGet1(10));
printTuple{int, string}(obj);
tupleVar = obj;

class t{V} <- tuple{int, V}
  def self(a::int, b::V) : super(a, b) = end
  def getV() -> V = self.second; end
end

def makeTuple{A, B}(a::A, b::B) -> tuple{A, B} = 
  tuple{A, B}(a, b); 
end

x = t{int}(10, 10);
printTuple{int, int}(x);
z = makeTuple{int, int}(10, 20);
printTuple{int, int}(z);
