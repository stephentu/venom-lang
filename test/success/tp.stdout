20
10
foo
10
10
10
20
